File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/distributions/distributions.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.spark.sql.connector.distributions

import org.apache.spark.sql.connector.expressions.{Expression, SortOrder}
import org.apache.spark.util.ArrayImplicits._

private[sql] object LogicalDistributions {

  def unspecified(): UnspecifiedDistribution = {
    UnspecifiedDistributionImpl
  }

  def clustered(clustering: Array[Expression]): ClusteredDistribution = {
    ClusteredDistributionImpl(clustering.toImmutableArraySeq)
  }

  def ordered(ordering: Array[SortOrder]): OrderedDistribution = {
    OrderedDistributionImpl(ordering.toImmutableArraySeq)
  }
}

private[sql] object UnspecifiedDistributionImpl extends UnspecifiedDistribution {
  override def toString: String = "UnspecifiedDistribution"
}

private[sql] final case class ClusteredDistributionImpl(
    clusteringExprs: Seq[Expression]) extends ClusteredDistribution {

  override def clustering: Array[Expression] = clusteringExprs.toArray

  override def toString: String = {
    s"ClusteredDistribution(${clusteringExprs.map(_.describe).mkString(", ")})"
  }
}

private[sql] final case class OrderedDistributionImpl(
    orderingExprs: Seq[SortOrder]) extends OrderedDistribution {

  override def ordering: Array[SortOrder] = orderingExprs.toArray

  override def toString: String = {
    s"OrderedDistribution(${orderingExprs.map(_.describe).mkString(", ")})"
  }
}

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/distributions/Distributions.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.spark.sql.connector.distributions;

import org.apache.spark.annotation.Experimental;
import org.apache.spark.sql.connector.expressions.Expression;
import org.apache.spark.sql.connector.expressions.SortOrder;

/**
 * Helper methods to create distributions to pass into Spark.
 *
 * @since 3.2.0
 */
@Experimental
public class Distributions {
  private Distributions() {
  }

  /**
   * Creates a distribution where no promises are made about co-location of data.
   */
  public static UnspecifiedDistribution unspecified() {
    return LogicalDistributions.unspecified();
  }

  /**
   * Creates a distribution where tuples that share the same values for clustering expressions are
   * co-located in the same partition.
   */
  public static ClusteredDistribution clustered(Expression[] clustering) {
    return LogicalDistributions.clustered(clustering);
  }

  /**
   * Creates a distribution where tuples have been ordered across partitions according
   * to ordering expressions, but not necessarily within a given partition.
   */
  public static OrderedDistribution ordered(SortOrder[] ordering) {
    return LogicalDistributions.ordered(ordering);
  }
}
